### PR TITLE
Switch to shiny new fast, RLP-less trie

### LINKED ISSUE #664: Exorcise RLP from codebase
Trie should be using `BareCodec` (which just uses `substrate_codec` to encode trie nodes), rather than `RlpCodec`.

### PR DESCRIPTION
**Pull request by gavofyork**
Closes #664 

Also expunges RLP from the codebase. There's not really any new logic outside of the `substrate-trie` crate.

The codebase is no longer generic over `NodeCodec`; the reason for this twofold: firstly, trie node format allows for much less opinionated selection than the hashing algorithm. Formats do not significantly change in terms of security. Performance tweaks are non-controversial and are easily folded in over time. Secondly, it would require a two generic parameters, not one as before.

### CODE REVIEW

File: core/state-machine/src/changes_trie/build.rs [hunk 1]
```diff
@@ -242,23 +239,23 @@ mod test {
...
 	fn build_changes_trie_nodes_on_digest_block_l1() {
 		let (backend, storage, changes) = prepare_for_build();
-		let changes_trie_nodes = prepare_input::<_, _, _, RlpCodec>(&backend, Some(&storage), &changes, 4).unwrap();
+		let changes_trie_nodes = prepare_input(&backend, Some(&storage), &changes, 4).unwrap();
 		assert_eq!(changes_trie_nodes, Some(vec![
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![100] }, vec![0, 2, 3]),
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![101] }, vec![1]),
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![103] }, vec![0, 1]),
 
-			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![100] }, vec![1, 3]),
```
Comment (gavofyork) [inline core/state-machine/src/changes_trie/build.rs]: @svyatonik I had to change this to make the tests work. any idea why?

File: core/state-machine/src/changes_trie/build.rs [hunk 2]
```diff
@@ -242,23 +239,23 @@ mod test {
...
+			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![100] }, vec![1]),
 			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![101] }, vec![1]),
 			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![102] }, vec![2]),
-			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![105] }, vec![1, 3]),
```
Comment (gavofyork) [inline core/state-machine/src/changes_trie/build.rs]: @svyatonik here, too

File: core/state-machine/src/changes_trie/build.rs [hunk 3]
```diff
@@ -282,16 +279,16 @@ mod test {
...
 
-		let changes_trie_nodes = prepare_input::<_, _, _, RlpCodec>(&backend, Some(&storage), &changes, 4).unwrap();
+		let changes_trie_nodes = prepare_input(&backend, Some(&storage), &changes, 4).unwrap();
 		assert_eq!(changes_trie_nodes, Some(vec![
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![100] }, vec![0, 2, 3]),
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![101] }, vec![1]),
 			InputPair::ExtrinsicIndex(ExtrinsicIndex { block: 4, key: vec![103] }, vec![0, 1]),
 
-			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![100] }, vec![1, 3]),
+			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![100] }, vec![1]),
```
Comment (gavofyork) [inline core/state-machine/src/changes_trie/build.rs]: @svyatonik ...and here...

File: core/state-machine/src/changes_trie/build.rs [hunk 4]
```diff
@@ -282,16 +279,16 @@ mod test {
...
 			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![101] }, vec![1]),
 			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![102] }, vec![2]),
-			InputPair::DigestIndex(DigestIndex { block: 4, key: vec![105] }, vec![1, 3]),
```
Comment (gavofyork) [inline core/state-machine/src/changes_trie/build.rs]: @svyatonik ...and here

**Comment by gavofyork** [comment]
@svyatonik quite a few of the change-trie tests include raw roots. there's little i can do except to replace with the expected root. i don't know how to ensure these trie changes haven't broken their logic.

**Comment by svyatonik** [comment]
@gavofyork The same is true for `Header::state_root` - it always changes when storage is altered => `https://github.com/paritytech/substrate/blob/master/node/executor/src/lib.rs` is updated often too. I'll take a look at the reason of "I had to change this to make the tests work. any idea why?", but generally I assume that the same steps should be taken with `changes_trie_root`.

**Comment by gavofyork** [comment]
@svyatonik yeah the hashes i'm less bothered about (though it is rather tedious). however, those four changes i pointed to above are switching out some of the structure's values, which is concerning.

File: core/state-machine/src/backend.rs [hunk 5]
```diff
@@ -179,24 +177,24 @@ impl<H: Hasher, C: NodeCodec<H>> Backend<H, C> for InMemory<H, C> where H::Out:
...
 		H: Hasher,
 		H::Out: HeapSizeOf,
-		C: NodeCodec<H>,
+	
 		I: Iterator<Item=(Vec<u8>, Vec<u8>)>,
 {
 	let mut root = <H as Hasher>::Out::default();
 	{
-		let mut trie = TrieDBMut::<H, C>::new(mdb, &mut root);
+		let mut trie = TrieDBMut::<H>::new(mdb, &mut root);
```
Comment (rphmeier) [inline core/state-machine/src/backend.rs]: did `TrieDBMut` lose a generic param?

File: core/trie/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,635 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
```
Comment (rphmeier) [inline core/trie/src/lib.rs]: docs are missing in this file/crate

File: core/trie/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+	let mut cursor: u16 = 1;
+	for v in has_children {
+		if v { bitmap |= cursor }
+		cursor <<= 1;
+	}
+	[first, (bitmap % 256 ) as u8, (bitmap / 256 ) as u8]
+}
+
+/*
+// TODO: Remove
```
Comment (rphmeier) [inline core/trie/src/lib.rs]: committing commented code

File: core/trie/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+	use super::*;
+	use codec::{Encode, Compact};
+	use substrate_primitives::Blake2Hasher;
+	use memory_db::MemoryDB;
+	use hash_db::{HashDB, Hasher};
+	use trie_db::{DBValue, TrieMut, Trie};
+	use trie_standardmap::{Alphabet, ValueMode, StandardMap};
+
+	fn check_equivalent(input: &Vec<(&[u8], &[u8])>) {
+/*		{
```
Comment (rphmeier) [inline core/trie/src/lib.rs]: here too

File: core/state-machine/src/changes_trie/build.rs (hunk 1 above)
Comment (gavofyork) [reply core/state-machine/src/changes_trie/build.rs]: ok if it's an issue with the iterator, then it would be good to have a test where i can reproduce it...

Comment (svyatonik) [reply core/state-machine/src/changes_trie/build.rs]: here's the test: https://github.com/paritytech/substrate/commit/93e13a8e1ce54522c2af7eb85c7348a92828fdd6

File: core/executor/Cargo.toml [hunk 9]
```diff
@@ -8,19 +8,20 @@ error-chain = "0.12"
...
 wasmi = "0.4"
 byteorder = "1.1"
-triehash = "0.2"
+trie-root = { git = "https://github.com/paritytech/trie" }
 twox-hash = "1.1.0"
 lazy_static = "1.0"
 parking_lot = "*"
 log = "0.3"
-hashdb = "0.2.1"
+hash-db = { git = "https://github.com/paritytech/trie" }
```
Comment (rphmeier) [inline core/executor/Cargo.toml]: kind of unfortunate that we are moving back off of crates.io deps and into GitHub land

File: core/trie/src/node_codec.rs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	fn empty_node() -> Vec<u8> {
+		vec![EMPTY_TRIE]
+	}
+
+	// TODO: refactor this so that `partial` isn't already encoded with HPE. Should just be an `impl Iterator<Item=u8>`.
+	fn leaf_node(partial: &[u8], value: &[u8]) -> Vec<u8> {
+		let mut output = partial_to_key(partial, LEAF_NODE_OFFSET, LEAF_NODE_BIG);
+		value.encode_to(&mut output);
+//		println!("leaf_node: {:#x?}", output);
```
Comment (rphmeier) [inline core/trie/src/node_codec.rs]: stray println commented here

File: core/trie/src/node_codec.rs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	// TODO: refactor this so that `partial` isn't already encoded with HPE. Should just be an `impl Iterator<Item=u8>`.
+	fn ext_node(partial: &[u8], child: ChildReference<H::Out>) -> Vec<u8> {
+		let mut output = partial_to_key(partial, EXTENSION_NODE_OFFSET, EXTENSION_NODE_BIG);
+		match child {
+			ChildReference::Hash(h) => 
+				h.as_ref().encode_to(&mut output),
+			ChildReference::Inline(inline_data, len) =>
+				(&AsRef::<[u8]>::as_ref(&inline_data)[..len]).encode_to(&mut output),
+		};
+//		println!("ext_node: {:#x?}", output);
```
Comment (rphmeier) [inline core/trie/src/node_codec.rs]: and here

File: core/trie/src/node_codec.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+				true
+			}
+			Some(ChildReference::Inline(inline_data, len)) => {
+				(&AsRef::<[u8]>::as_ref(&inline_data)[..len]).encode_to(&mut output);
+				true
+			}
+			None => false,
+		}));
+		output[0..3].copy_from_slice(&prefix[..]);
+//		println!("branch_node: {:#x?}", output);
```
Comment (rphmeier) [inline core/trie/src/node_codec.rs]: and here

File: core/trie/src/node_codec_alt.rs [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! `NodeCodec` implementation for Rlp
```
Comment (rphmeier) [inline core/trie/src/node_codec_alt.rs]: is this "alt" implementation for RLP for Parity Codec?

File: core/state-machine/src/changes_trie/build.rs (hunk 1 above)
Comment (gavofyork) [reply core/state-machine/src/changes_trie/build.rs]: thanks. that's fixed now, but the fix didn't invalidate the changes here...

File: core/executor/Cargo.toml (hunk 9 above)
Comment (gavofyork) [reply core/executor/Cargo.toml]: once it stablises (i.e. sometime between now and 1.0 final), i'm happy to publish and move back.

File: core/state-machine/src/changes_trie/build.rs (hunk 1 above)
Comment (gavofyork) [reply core/state-machine/src/changes_trie/build.rs]: nm false alarm. tests confirm fix.

File: core/trie/README.md [hunk 14]
```diff
@@ -0,0 +1,2 @@
+This crate provides utility functions to validate and initialize tries using flexible input.
+It is used extensively in `parity-ethereum` to validate blocks (mostly transactions and receipt roots).
```
Comment (arkpar) [inline core/trie/README.md]: This looks irrelevant.

File: core/trie/src/node_header.rs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	EXTENSION_NODE_BIG, BRANCH_NODE_NO_VALUE, BRANCH_NODE_WITH_VALUE, LEAF_NODE_THRESHOLD,
+	EXTENSION_NODE_THRESHOLD, LEAF_NODE_SMALL_MAX, EXTENSION_NODE_SMALL_MAX};
+
+/// A node header.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum NodeHeader {
+	Null,
+	Branch(bool),
+	Extension(usize),
+	Leaf(usize),
```
Comment (arkpar) [inline core/trie/src/node_header.rs]: usize? Is it supposed to be platform-dependent? `u32` should probably enough

File: core/trie/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,421 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
```
Comment (arkpar) [inline core/trie/src/lib.rs]: Needs updating?

File: core/trie/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+pub fn trie_root<H: Hasher, I, A, B>(input: I) -> H::Out where
+	I: IntoIterator<Item = (A, B)>,
+	A: AsRef<[u8]> + Ord,
+	B: AsRef<[u8]>,
+{
+	trie_root::trie_root::<H, TrieStream, _, _, _>(input)
+}
+
+/// Determine a trie root node's data given its ordered contents, closed form.
+pub fn unhashed_trie<H: Hasher, I, A, B>(input: I) -> Vec<u8> where
```
Comment (dvdplm) [inline core/trie/src/lib.rs]: Is this method used for anything outside tests?

File: core/trie/src/node_codec.rs [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+	EXTENSION_NODE_BIG, take, partial_to_key, node_header::NodeHeader, branch_node};
+
+/// Concrete implementation of a `NodeCodec` with Parity Codec encoding, generic over the `Hasher`
+#[derive(Default, Clone)]
+pub struct NodeCodec<H: Hasher>(PhantomData<H>);
+
+// NOTE: what we'd really like here is:
+// `impl<H: Hasher> NodeCodec<H> for RlpNodeCodec<H> where H::Out: Decodable`
+// but due to the current limitations of Rust const evaluation we can't
+// do `const HASHED_NULL_NODE: H::Out = H::Out( … … )`. Perhaps one day soon?
```
Comment (dvdplm) [inline core/trie/src/node_codec.rs]: This comment is probably not helpful anymore.

File: core/trie/src/trie_stream.rs [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+//		println!("[append_substream] START other buffer: {:x?}", data);
+		match data.len() {
+			0...31 => {
+//				println!("[append_substream] appending data, because data.len() = {}", data.len());
+				data.encode_to(&mut self.buffer)
+			},
+			_ => {
+//				println!("[append_substream] would have hashed, because data.len() = {}", data.len());
+//				data.encode_to(&mut self.buffer)
+				// TODO: re-enable hashing before merging
```
Comment (dvdplm) [inline core/trie/src/trie_stream.rs]: Some cleanup to do in this file.
